### PR TITLE
feat(nbstore): add awareness storage&sync&frontend

### PR DESCRIPTION
**Pull request by EYHN**

### CODE REVIEW

**Comment by EYHN** [comment]
* **#9016** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9016?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#9110](https://github.com/toeverything/AFFiNE/pull/9110) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9110?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#9131](https://github.com/toeverything/AFFiNE/pull/9131) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9131?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9016?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#9084** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9084?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9070** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9070?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#9037** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/9037?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8996** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8996?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8918** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8918?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/toeverything?ref=merge-queue-instructions-in-stack-comment&prId=9016).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/common/nbstore/src/impls/idb/awareness.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import { DummyConnection } from '../../connection';
+import {
+  type AwarenessRecord,
+  AwarenessStorage,
+} from '../../storage/awareness';
+
+export class IndexedDBAwarenessStorage extends AwarenessStorage {
```
Comment (forehalo) [inline packages/common/nbstore/src/impls/idb/awareness.ts]: BroadcastChannel

File: packages/common/nbstore/src/impls/broadcast-channel/awareness.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+      bin: record.bin,
+      origin,
+    } satisfies ChannelMessage);
+    return Promise.resolve();
+  }
+
+  override subscribeUpdate(
+    id: string,
+    onUpdate: (update: AwarenessRecord, origin?: string) => void,
+    onCollect: () => AwarenessRecord
```
Comment (forehalo) [inline packages/common/nbstore/src/impls/broadcast-channel/awareness.ts:60]: always store last awareness and auto send when others collect?

Comment (EYHN) [reply packages/common/nbstore/src/impls/broadcast-channel/awareness.ts:60]: awareness only transmit the changed parts each time.

Comment (EYHN) [reply packages/common/nbstore/src/impls/broadcast-channel/awareness.ts:60]: In onCollect, each client will only send its own part of the status, which can also be used to filter offline clients.

**Review by forehalo — APPROVED** [review]

**Pull request merged**